File: lib/core/domain/guided/guided_composer.dart
import 'package:lsb_legal_app/core/domain/guided/guided_answer.dart';
import 'package:lsb_legal_app/core/domain/guided/question_bank.dart';

/// Texto redactado y qué respuestas quedaron representadas en él.
class ComposedText {
  final String text;

  /// Opciones cuya frase entró en [text], como `'Q.ID#opcion'`.
  final Set<String> represented;

  const ComposedText(this.text, this.represented);
}

/// Redacción determinista de una intervención guiada.
///
/// Gemelo exacto de `aws/guided_composer.py`: los dos leen el mismo banco y
/// deben producir la misma oración para las mismas respuestas. Nada aquí
/// interpreta ni completa: cada oración sale de la plantilla de la opción
/// elegida y de los valores escritos por la persona.
class GuidedComposer {
  final QuestionBank bank;

  const GuidedComposer(this.bank);

  String compose(GuidedIntervention intervention) =>
      composeTraced(intervention).text;

  /// Hechos confirmados: opciones elegidas (no omitidas) que redactan algo.
  ///
  /// Las opciones que no redactan nada por sí mismas solo abren otra
  /// pregunta (p. ej. «Alguien escapó» → «¿Quién escapó?»), que es la que
  /// lleva el hecho; por eso no cuentan aquí.
  Set<String> confirmedFacts(GuidedIntervention intervention) => {
        for (final a in intervention.answers)
          if (!a.isOmitted)
            for (final id in a.optionIds)
              if (bank.question(a.questionId)?.option(id) case final option?)
                if (option.writesSomething) '${a.questionId}#$id',
      };

  ComposedText composeTraced(GuidedIntervention intervention) {
    final answers = <String, Map<String, dynamic>>{
      for (final answer in intervention.answers)
        answer.questionId: answer.toJson(),
    };
    final consumed = <String>{};
    final represented = <String>{};
    final sentences = <String>[];
    final reply = intervention.purpose == GuidedPurpose.reply;
    final referenced = _referencedFragments(answers);

    String? fragment(String id, List<String> extras) {
      final answer = answers[id];
      final question = bank.questions[id];
      if (answer == null || question == null || answer['estado'] == 'omitido') {
        return null;
      }
      consumed.add(id);
      final parts = <String>[];
      for (final option in _chosen(question, answer)) {
        final phrase = _fill(
            option, _phraseOf(option, answer), answer, answers, fragment, extras);
        final extra = option['fraseExtra'] as String?;
        if (extra != null) extras.add(extra);
        if (phrase.isNotEmpty) {
          parts.add(phrase);
          represented.add('$id#${option['id']}');
        }
      }
      return parts.isEmpty ? null : _joinEs(parts);
    }

    for (final id in _order(intervention, answers)) {
      if (consumed.contains(id)) continue;
      final answer = answers[id];
      final question = bank.questions[id];
      if (answer == null || question == null || answer['estado'] == 'omitido') {
        continue;
      }
      final extras = <String>[];
      // Lo que se marque al redactar esta frase solo cuenta si la frase
      // llega al texto.
      final before = {...represented};
      String text;
      if (question['modo'] == 'fragmento') {
        // Lo redacta la pregunta que lo cita («Me robaron {Q.ROB.QUE}»); solo
        // si ninguna lo cita se usa su frase suelta.
        if (referenced.contains(id)) continue;
        final value = fragment(id, extras);
        final loose = question['fraseSuelta'] as String?;
        text = value == null || loose == null ? '' : loose.replaceAll('{frag}', value);
      } else {
        consumed.add(id);
        final chosen = _chosen(question, answer);
        final parts = <String>[];
        for (final option in chosen) {
          final result = _fill(option, _phraseOf(option, answer), answer,
              answers, fragment, extras);
          final extra = option['fraseExtra'] as String?;
          if (extra != null) extras.add(extra);
          if (result.isNotEmpty) {
            parts.add(result);
            represented.add('$id#${option['id']}');
          }
        }
        final hasExit = chosen.any((option) => option['salida'] == true);
        if (parts.isEmpty) {
          text = '';
        } else if (question['plantilla'] != null && !hasExit) {
          final joined = _joinEs(parts, spaces: question['unir'] == 'espacio');
          text = (question['plantilla'] as String).replaceAll('{items}', joined);
        } else {
          text = parts.join(' ');
        }
      }
      final finalized = _finalize(text, reply: reply);
      if (finalized.isNotEmpty) {
        sentences.add(finalized);
      } else {
        represented
          ..clear()
          ..addAll(before);
      }
      sentences.addAll(extras
          .map((e) => _finalize(e, reply: reply))
          .where((e) => e.isNotEmpty));
    }
    return ComposedText(sentences.join(' '), represented);
  }

  /// Preguntas en modo fragmento que alguna respuesta elegida cita
  /// explícitamente (`{Q.ID|respaldo}`) y que, por tanto, se redactan dentro
  /// de la frase de quien las cita.
  Set<String> _referencedFragments(Map<String, Map<String, dynamic>> answers) {
    final out = <String>{};
    for (final entry in answers.entries) {
      final answer = entry.value;
      final question = bank.questions[entry.key];
      if (question == null || answer['estado'] == 'omitido') continue;
      for (final option in _chosen(question, answer)) {
        final template = _phraseOf(option, answer);
        for (final match in RegExp(r'\{([^{}]+)\}').allMatches(template)) {
          final token = match.group(1)!;
          if (!token.startsWith('Q.') && !token.startsWith('I.')) continue;
          final separator = token.indexOf('|');
          final reference = separator < 0 ? token : token.substring(0, separator);
          final resolved = _resolveReference(reference);
          if (resolved == null || resolved.$2.isNotEmpty) continue;
          final target = bank.questions[resolved.$1];
          final targetAnswer = answers[resolved.$1];
          if (target?['modo'] == 'fragmento' &&
              targetAnswer != null &&
              targetAnswer['estado'] != 'omitido') {
            out.add(resolved.$1);
          }
        }
      }
    }
    return out;
  }

  List<String> _order(GuidedIntervention intervention,
      Map<String, Map<String, dynamic>> answers) {
    final result = <String>[];
    final journeys = bank.data['recorridos'] as Map<String, dynamic>? ?? const {};
    final journey = journeys[intervention.journeyId] as Map<String, dynamic>?;
    final base = journey == null
        ? intervention.steps
        : ((journey['ordenRedaccion'] as List<dynamic>?) ??
                (journey['pasos'] as List<dynamic>).map(
                    (step) => (step as Map<String, dynamic>)['pregunta']))
            .cast<String>();
    for (final id in [...base, ...answers.keys]) {
      if (!result.contains(id)) result.add(id);
    }
    return result;
  }

  List<Map<String, dynamic>> _chosen(
      Map<String, dynamic> question, Map<String, dynamic> answer) {
    final ids = (answer['opciones'] as List<dynamic>? ?? const []).toSet();
    return (question['opciones'] as List<dynamic>? ?? const [])
        .cast<Map<String, dynamic>>()
        .where((option) => ids.contains(option['id']))
        .toList();
  }

  Map<String, dynamic> _values(
      Map<String, dynamic> answer, Map<String, dynamic> option) {
    final all = answer['valores'] as Map<String, dynamic>? ?? const {};
    return all[option['id']] as Map<String, dynamic>? ?? const {};
  }

  String _phraseOf(Map<String, dynamic> option, Map<String, dynamic> answer) {
    final values = _values(answer, option);
    if (option['fraseSingular'] != null && '${values['n'] ?? ''}' == '1') {
      return option['fraseSingular'] as String;
    }
    final editor = option['editor'] as String?;
    final keys = _editorKeys[editor] ?? const <String>[];
    final complete = keys.isNotEmpty &&
        keys.every((key) => '${values[key] ?? ''}'.trim().isNotEmpty);
    if (editor != null && !complete && option.containsKey('fraseSinValor')) {
      return option['fraseSinValor'] as String? ?? '';
    }
    return option['frase'] as String? ?? '';
  }

  String _fill(
      Map<String, dynamic> option,
      String template,
      Map<String, dynamic> answer,
      Map<String, Map<String, dynamic>> answers,
      String? Function(String, List<String>) fragment,
      List<String> extras) {
    final values = _values(answer, option);
    final mention = answer['mencion'] as Map<String, dynamic>?;
    return template.replaceAllMapped(RegExp(r'\{([^{}]+)\}'), (match) {
      final token = match.group(1)!;
      if (token.startsWith('Q.') || token.startsWith('I.')) {
        final separator = token.indexOf('|');
        final reference = separator < 0 ? token : token.substring(0, separator);
        final fallback = separator < 0 ? '' : token.substring(separator + 1);
        final resolved = _resolveReference(reference);
        if (resolved == null) return fallback;
        final referenced = answers[resolved.$1];
        if (resolved.$2.isNotEmpty) {
          if (referenced == null || referenced['estado'] != 'afirmado') return fallback;
          for (final selected in _chosen(bank.questions[resolved.$1]!, referenced)) {
            final value = selected[resolved.$2];
            if (value != null) return '$value';
          }
          return fallback;
        }
        if (referenced == null || referenced['estado'] == 'omitido') return fallback;
        return fragment(resolved.$1, extras) ?? fallback;
      }
      if (token == 'aprox') return values['aprox'] == 'si' ? 'aproximadamente ' : '';
      if (token == 'mencionado') return '${mention?['frase'] ?? 'eso'}';
      if (token.startsWith('?')) {
        final value = '${values[token.substring(1)] ?? ''}'.trim();
        return value.isEmpty ? '' : ' $value';
      }
      return '${values[token] ?? ''}'.trim();
    });
  }

  (String, String)? _resolveReference(String reference) {
    final parts = reference.split('.');
    for (var length = parts.length; length > 1; length--) {
      final id = parts.take(length).join('.');
      if (bank.questions.containsKey(id)) {
        return (id, parts.skip(length).join('.'));
      }
    }
    return null;
  }

  String _finalize(String text, {required bool reply}) {
    var result = text
        .replaceAll(RegExp(r'\s+'), ' ')
        .replaceAll(RegExp(r'\s+([.,;:?!»])'), r'$1')
        .replaceAll(RegExp(r'«\s+'), '«')
        .trim();
    if (result.isEmpty) return '';
    if (!reply) {
      for (final prefix in const ['Sí, ', 'No, ']) {
        if (result.startsWith(prefix)) result = result.substring(prefix.length);
      }
    }
    result = result[0].toUpperCase() + result.substring(1);
    if (!'.?!'.contains(result[result.length - 1])) result += '.';
    return result;
  }
}

const _editorKeys = <String, List<String>>{
  'texto_nombre': ['nombre'],
  'telefono': ['telefono'],
  'entero': ['n'],
  'edad': ['n'],
  'monto': ['monto', 'moneda'],
  'texto_detalle': ['texto'],
  'lugar_literal': ['nombre'],
  'referencia': ['referencia'],
  'documento_numero': ['numero'],
  'hora': ['hora'],
};

String _joinEs(List<String> items, {bool spaces = false}) {
  if (spaces) return items.join(' ');
  if (items.length < 2) return items.join();
  return '${items.take(items.length - 1).join(', ')} y ${items.last}';
}

File: aws/guided_composer.py
"""Redacción de una intervención guiada a partir de respuestas tipadas.

Gemelo exacto de `lib/core/domain/guided/guided_composer.dart`. Los dos leen
el mismo banco (`question_bank.json`, generado por
`tool/build_question_matrix.py` desde `docs/negocio/config/banco_preguntas.json`)
y deben producir la misma oración para las mismas respuestas: lo comprueban
los fixtures que escribe el cliente (`test/fixtures/guided/`) y que
`aws/tests/test_guiado_v4.py` pasa por el handler.

Nada aquí interpreta ni completa: cada oración sale de la plantilla de la
opción elegida y de los valores escritos por la persona. Si una respuesta no
está, no se redacta; si es «no sé», se redacta «no sé».
"""

from __future__ import annotations

import json
import os
import re

BANK_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "question_bank.json")

_TOKEN = re.compile(r"\{([^{}]+)\}")

# Qué clave de valor rellena cada editor.
EDITOR_KEYS = {
    "texto_nombre": ("nombre",),
    "telefono": ("telefono",),
    "entero": ("n",),
    "edad": ("n",),
    "monto": ("monto", "moneda"),
    "texto_detalle": ("texto",),
    "lugar_literal": ("nombre",),
    "referencia": ("referencia",),
    "documento_numero": ("numero",),
    "hora": ("hora",),
}

ESTADOS = ("afirmado", "negado", "desconocido", "omitido")
PROPOSITOS = ("standalone", "initiative", "reply")


def load_bank(path: str = BANK_PATH) -> dict:
    with open(path, encoding="utf-8") as f:
        return json.load(f)


def join_es(items: list, espacio: bool = False) -> str:
    items = [i for i in items if i]
    if espacio:
        return " ".join(items)
    if not items:
        return ""
    if len(items) == 1:
        return items[0]
    return ", ".join(items[:-1]) + " y " + items[-1]


def _normalize(text: str) -> str:
    text = re.sub(r"\s+", " ", text).strip()
    text = re.sub(r"\s+([.,;:?!»])", r"\1", text)
    text = re.sub(r"«\s+", "«", text)
    return text


def _capitalize(text: str) -> str:
    if not text:
        return text
    return text[0].upper() + text[1:]


class Composer:
    def __init__(self, bank: dict):
        self.bank = bank
        self.questions = {q["id"]: q for q in bank["preguntas"]}

    # ------------------------------------------------------------------
    def compose(self, guided: dict) -> str:
        return self.compose_traced(guided)[0]

    def confirmed_facts(self, guided: dict) -> set:
        """Opciones elegidas (no omitidas) que redactan algo, `'Q.ID#opcion'`.

        Gemelo de `GuidedComposer.confirmedFacts`. Las opciones que no
        redactan nada solo abren la pregunta que lleva el hecho.
        """
        out = set()
        for a in guided.get("respuestas") or []:
            if a.get("estado") == "omitido":
                continue
            q = self.questions.get(a.get("pregunta"))
            if q is None:
                continue
            for o in self._chosen(q, a):
                if _writes_something(o):
                    out.add(f"{a['pregunta']}#{o['id']}")
        return out

    def compose_traced(self, guided: dict) -> tuple:
        """(texto, opciones representadas). Gemelo de `composeTraced`."""
        self._answers = {}
        for a in guided.get("respuestas") or []:
            self._answers[a["pregunta"]] = a
        self._consumed = set()
        self._represented = set()
        self._reply = guided.get("proposito") == "reply"
        referenced = self._referenced_fragments()
        sentences = []

        for qid in self._order(guided):
            if qid in self._consumed:
                continue
            a = self._answers.get(qid)
            if not a or a.get("estado") == "omitido":
                continue
            q = self.questions.get(qid)
            if q is None:
                continue
            self._extras = []
            # Lo que se marque al redactar esta frase solo cuenta si la frase
            # llega al texto.
            antes = set(self._represented)
            if q.get("modo") == "fragmento":
                # Lo redacta la pregunta que lo cita; solo si ninguna lo cita
                # se usa su frase suelta.
                if qid in referenced:
                    continue
                frag = self._fragment(qid)
                if frag is None or not q.get("fraseSuelta"):
                    self._represented = antes
                    sentences.extend(self._finalize(e) for e in self._extras)
                    continue
                text = q["fraseSuelta"].replace("{frag}", frag)
            else:
                self._consumed.add(qid)
                text = self._sentence(q, a)
            text = self._finalize(text)
            if text:
                sentences.append(text)
            else:
                self._represented = antes
            sentences.extend(self._finalize(e) for e in self._extras)
        return " ".join(s for s in sentences if s), set(self._represented)

    def _referenced_fragments(self) -> set:
        """Preguntas en modo fragmento citadas por alguna respuesta elegida."""
        out = set()
        for qid, a in self._answers.items():
            q = self.questions.get(qid)
            if q is None or a.get("estado") == "omitido":
                continue
            for o in self._chosen(q, a):
                for token in _TOKEN.findall(self._phrase_of(o, a)):
                    if not token.startswith(("Q.", "I.")):
                        continue
                    ref = token.partition("|")[0]
                    target, atributo = self._resolve_ref(ref)
                    if target is None or atributo:
                        continue
                    ra = self._answers.get(target)
                    if (self.questions[target].get("modo") == "fragmento"
                            and ra and ra.get("estado") != "omitido"):
                        out.add(target)
        return out

    # ------------------------------------------------------------------
    def _order(self, guided: dict) -> list:
        orden = []
        rec = self.bank.get("recorridos", {}).get(guided.get("recorrido") or "")
        if rec:
            base = rec.get("ordenRedaccion") or [p["pregunta"] for p in rec["pasos"]]
        else:
            base = list(guided.get("pasos") or [])
        for qid in base:
            if qid not in orden:
                orden.append(qid)
        for a in guided.get("respuestas") or []:
            if a["pregunta"] not in orden:
                orden.append(a["pregunta"])
        return orden

    def _chosen(self, q: dict, a: dict) -> list:
        elegidas = set(a.get("opciones") or [])
        return [o for o in q.get("opciones", []) if o["id"] in elegidas]

    def _values(self, a: dict, o: dict) -> dict:
        return ((a.get("valores") or {}).get(o["id"])) or {}

    def _has_values(self, o: dict, vals: dict) -> bool:
        claves = EDITOR_KEYS.get(o.get("editor") or "", ())
        return bool(claves) and all(str(vals.get(k, "")).strip() for k in claves)

    def _phrase_of(self, o: dict, a: dict) -> str:
        vals = self._values(a, o)
        if o.get("editor"):
            if o.get("fraseSingular") and str(vals.get("n", "")) == "1":
                return o["fraseSingular"]
            if not self._has_values(o, vals) and o.get("fraseSinValor") is not None:
                return o["fraseSinValor"]
        return o.get("frase") or ""

    def _sentence(self, q: dict, a: dict) -> str:
        chosen = self._chosen(q, a)
        if not chosen:
            return ""
        salida = any(o.get("salida") for o in chosen)
        partes = []
        for o in chosen:
            texto = self._fill(o, self._phrase_of(o, a), a)
            if texto:
                partes.append(texto)
                self._represented.add(f"{a['pregunta']}#{o['id']}")
            if o.get("fraseExtra"):
                self._extras.append(o["fraseExtra"])
        if not partes:
            return ""
        if q.get("plantilla") and not salida:
            unidas = join_es(partes, espacio=q.get("unir") == "espacio")
            return q["plantilla"].replace("{items}", unidas)
        return " ".join(partes)

    def _fragment(self, qid: str):
        a = self._answers.get(qid)
        if not a or a.get("estado") == "omitido":
            return None
        q = self.questions.get(qid)
        if q is None:
            return None
        self._consumed.add(qid)
        partes = []
        for o in self._chosen(q, a):
            texto = self._fill(o, self._phrase_of(o, a), a)
            if texto:
                partes.append(texto)
                self._represented.add(f"{qid}#{o['id']}")
            if o.get("fraseExtra"):
                self._extras.append(o["fraseExtra"])
        if not partes:
            return None
        return join_es(partes)

    def _resolve_ref(self, ref: str):
        partes = ref.split(".")
        for n in range(len(partes), 1, -1):
            qid = ".".join(partes[:n])
            if qid in self.questions:
                return qid, ".".join(partes[n:])
        return None, None

    def _fill(self, o: dict, template: str, a: dict) -> str:
        vals = self._values(a, o)
        mencion = (a.get("mencion") or {}).get("frase") or ""

        def sustituir(m):
            token = m.group(1)
            if token.startswith(("Q.", "I.")):
                ref, _, respaldo = token.partition("|")
                qid, atributo = self._resolve_ref(ref)
                if qid is None:
                    return respaldo
                if atributo:
                    ra = self._answers.get(qid)
                    if not ra or ra.get("estado") != "afirmado":
                        return respaldo
                    for elegida in self._chosen(self.questions[qid], ra):
                        if elegida.get(atributo):
                            return elegida[atributo]
                    return respaldo
                ra = self._answers.get(qid)
                if not ra or ra.get("estado") in ("omitido",):
                    return respaldo
                frag = self._fragment(qid)
                return frag if frag else respaldo
            if token == "aprox":
                return "aproximadamente " if vals.get("aprox") == "si" else ""
            if token == "mencionado":
                return mencion or "eso"
            if token.startswith("?"):
                v = str(vals.get(token[1:], "")).strip()
                return " " + v if v else ""
            return str(vals.get(token, "")).strip()

        return _TOKEN.sub(sustituir, template)

    def _finalize(self, text: str) -> str:
        text = _normalize(text)
        if not text:
            return ""
        if not self._reply:
            for prefijo in ("Sí, ", "No, "):
                if text.startswith(prefijo):
                    text = text[len(prefijo):]
                    break
        text = _capitalize(text)
        if text[-1] not in ".?!":
            text += "."
        return text


def _writes_something(o: dict) -> bool:
    return any(str(o.get(k) or "").strip()
               for k in ("frase", "fraseSinValor", "fraseSingular"))


def compose(guided: dict, bank: dict = None) -> str:
    return Composer(bank or load_bank()).compose(guided)
